crates/ruma-signatures/src/keys.rs

```diff
@@ -42,27 +87,64 @@ impl Ed25519KeyPair {
     ///
     /// Returns an error if the public and private keys provided are invalid for the implementing
     /// algorithm.
-    pub fn new(document: &[u8], version: String) -> Result<Self, Error> {
-        let keypair = RingEd25519KeyPair::from_pkcs8(document)
-            .map_err(|error| Error::new(error.to_string()))?;
+    ///
+    /// Returns an error when the PKCS#8 document had a public key, but it doesn't match the one
+    /// generated from the private key. This is a fallback and extra validation against
+    /// corruption or
+    pub fn from_der(document: &[u8], version: String) -> Result<Self, Error> {
+        let oak = OneAsymmetricKey::from_der(document).map_err(|e| Error::new(format!("{}", e)))?;
+
+        Self::from_pkcs8_oak(oak, version)
+    }
+
+    /// Constructs a key pair from [`pkcs8::OneAsymmetricKey`].
+    pub fn from_pkcs8_oak(oak: OneAsymmetricKey<'_>, version: String) -> Result<Self, Error> {
+        Self::new(oak.algorithm.oid, oak.private_key, oak.public_key, version)
+    }
+
+    /// Constructs a key pair from [`pkcs8::PrivateKeyInfo`].
+    pub fn from_pkcs8_pki(oak: PrivateKeyInfo<'_>, version: String) -> Result<Self, Error> {
+        Self::new(oak.algorithm.oid, oak.private_key, None, version)
+    }
 
-        Ok(Self { keypair, version })
+    /// PKCS#8's "private key" is not yet actually the entire key,
+    /// so convert it if it is wrongly formatted.
+    ///
+    /// See [RFC 8310 10.3](https://datatracker.ietf.org/doc/html/rfc8410#section-10.3) for more details
+    fn correct_privkey_from_octolet(key: &[u8]) -> &[u8] {
+        if key.len() == 34 && key[..2] == [0x04, 0x20] {
+            &key[2..]
+        } else {
+            key
+        }
     }
 
     /// Generates a new key pair.
     ///
     /// # Returns
     ///
-    /// Returns a Vec<u8> representing a pkcs8-encoded private/public keypair
+    /// Returns a Vec<u8> representing a DER-encoded PKCS#8 v2 document (with public key)
     ///
     /// # Errors
     ///
     /// Returns an error if the generation failed.
     pub fn generate() -> Result<Vec<u8>, Error> {
-        let document = RingEd25519KeyPair::generate_pkcs8(&ring::rand::SystemRandom::new())
-            .map_err(|e| Error::new(e.to_string()))?;
+        let secret = SecretKey::generate(&mut rand::rngs::OsRng);
+
+        let public = Into::<PublicKey>::into(&secret);
```

Comment:
Would this work too?

```suggestion
        let public = PublicKey::from(&secret);
```

Otherwise:

```suggestion
        let public: PublicKey = (&secret).into();
```